satellite/satellitedb/orders.go

```diff
@@ -214,3 +218,155 @@ func (db *ordersDB) UnuseSerialNumber(ctx context.Context, serialNumber storj.Se
+				updateRollupStatement += db.buildUpdateStorageNodeBandwidthRollupStatements(request.OrderLimit, request.Order, intervalStart)
+			} else {
+				reject(request.OrderLimit.SerialNumber)
+			}
+		}
+	}
+
+	_, err = tx.Exec(updateRollupStatement)
+	if err != nil {
+		return []*pb.SettlementResponse{}, errs.Wrap(err)
+	}
+	for _, request := range requests {
+		_, rejected := rejectedRequests[request.OrderLimit.SerialNumber]
+		if !rejected {
+			r := &pb.SettlementResponse{
+				SerialNumber: request.OrderLimit.SerialNumber,
+				Status:       pb.SettlementResponse_ACCEPTED,
+			}
+
+			responses = append(responses, r)
+		}
+	}
+	return responses, nil
+}
+
+func (db *ordersDB) buildInsertUseSerialStatement(orderLimit *pb.OrderLimit) string {
+	return fmt.Sprintf("INSERT INTO  used_serials (serial_number_id, storage_node_id) SELECT id, %v FROM serial_numbers WHERE serial_number = %s;\n",
+		db.toHex(orderLimit.StorageNodeId.Bytes()), db.toHex(orderLimit.SerialNumber.Bytes()))
+}
+
+func (db *ordersDB) buildGetBucketIdsQuery(argCount int) string {
+	args := make([]string, argCount)
+	for i := 0; i < argCount; i++ {
+		args[i] = "?"
+	}
+	return fmt.Sprintf("SELECT serial_number, bucket_id FROM serial_numbers WHERE serial_number IN (%s);\n", strings.Join(args, ","))
+}
+
+func (db *ordersDB) buildUpdateBucketBandwidthRollupStatements(orderLimit *pb.OrderLimit, order *pb.Order, projectID []byte, bucketName []byte, intervalStart time.Time) string {
+	return fmt.Sprintf(`INSERT INTO bucket_bandwidth_rollups (bucket_name, project_id, interval_start, interval_seconds, action, inline, allocated, settled)
+		VALUES (%s, %s, '%s', %d, %d, %d, %d, %d)
+		ON CONFLICT(bucket_name, project_id, interval_start, action)
+		DO UPDATE SET settled = bucket_bandwidth_rollups.settled + %d;
+`, db.toHex(bucketName), db.toHex(projectID), intervalStart.Format("2006-01-02 15:04:05+00:00"), defaultIntervalSeconds, orderLimit.Action, 0, 0, uint64(order.Amount), uint64(order.Amount))
+}
+
+func (db *ordersDB) buildUpdateStorageNodeBandwidthRollupStatements(orderLimit *pb.OrderLimit, order *pb.Order, intervalStart time.Time) string {
+	return fmt.Sprintf(`INSERT INTO storagenode_bandwidth_rollups (storagenode_id, interval_start, interval_seconds, action, allocated, settled)
+		VALUES (%s, '%s', %d, %d, %d, %d)
+		ON CONFLICT(storagenode_id, interval_start, action)
+		DO UPDATE SET settled = storagenode_bandwidth_rollups.settled + %d;
+`, db.toHex(orderLimit.StorageNodeId.Bytes()), intervalStart.Format("2006-01-02 15:04:05+00:00"), defaultIntervalSeconds, orderLimit.Action, 0, uint64(order.Amount), uint64(order.Amount))
+}
+
+func (db *ordersDB) toHex(value []byte) string {
+	hexValue := hex.EncodeToString(value)
+	switch db.db.Driver().(type) {
+	case *sqlite3.SQLiteDriver:
+		return fmt.Sprintf("X'%v'", hexValue)
+	case *pq.Driver:
+		return fmt.Sprintf("decode('%v', 'hex')", hexValue)
+	default:
+		return ""
```

Comment:
This seems like a dangerous thing to me. I know that we use the string as a value in places, so it would be a lot of extra code checking errors, but I'd feel much more comfortable if adding a new database backend (which is expected to happen) didn't have to include finding this code silently corrupting the data (even in the best case of unit test failures).

Reply:
Updated